### DISCUSSION TITLE
/hyndsight/nnetar-prediction-intervals/

### DISCUSSION THREAD

**Discussion by robjhyndman** [robjhyndman/robjhyndman.com#273] [Announcements]
Prediction intervals for NNETAR models 
 https://robjhyndman.com/hyndsight/nnetar-prediction-intervals/

**Comment by robjhyndman** [comment]
```header
Originally posted by Felix on 2017-05-26 15:54:14
````
Hi Rob.

Please, i have a few question.

How can i control number of iteration in arima model with forecast
package?

Have you considered incorporating the sigma parameter as does the
eviews, including the other methods for estimating standard errors?

Thanks a lot for your time.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-05-27 03:40:44
````
I have no idea what you mean.

**Reply by robjhyndman** [reply]
```header
Originally posted by Felix on 2017-05-27 19:00:12
````
[https://uploads.disquscdn.c\...](https://uploads.disquscdn.com/images/6361048a643c38275e7e484d06d6f5f76cbb922ddd09d79ccd98ffc37c82dc9a.png)
In eviews I can put the maximum number of iterations and I can also
choose the estimation method, followed by the covariance estimation
method (for example: HAC)

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-05-28 06:39:18
````
The iterations are to do with the optimizer. You can control the
optimization of Arima in R using the optim.control argument. Read the
help file for stats::arima.

You can select the estimation method in R. Again, read the help file.
That\'s why we write them.

You cannot change the method for estimating the covariance matrix in R.
I\'m not sure why you would want to.

**Reply by robjhyndman** [reply]
```header
Originally posted by Felix on 2017-05-29 17:37:35
````
Thanks a lot!

**Comment by robjhyndman** [comment]
```header
Originally posted by cameronfen on 2017-05-27 22:54:28
````
You can build a Bayesian Neural Net if you want confidence intervals

**Comment by robjhyndman** [comment]
```header
Originally posted by Rohit on 2017-05-29 11:21:13
````
Hi Rob,

Thanks for the detailed blog.\
I am not getting prediction interval with PI = TRUE in forecast(fit,
xreg = xreg.val, PI=TRUE, h=n.val). I just get point forecast. Wondering
if this has anything to do with missing data in my time series?

Thanks,\
Rohit

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-05-29 23:23:29
````
The following works.

library(forecast)\
y \&lt;- ts(rnorm(50))\
x \&lt;- ts(rnorm(50))\
fit \&lt;- nnetar(y, xreg=x)\
forecast(fit, PI=TRUE, xreg=rnorm(4))

Missing values may cause problems, but well before computing prediction
intervals.

Make sure you are using the current version of the package (8.x)

**Comment by robjhyndman** [comment]
```header
Originally posted by pramod kishore on 2017-06-09 12:53:40
````
Hi Rob,

May be it is a very basic question.But please answer.\
My time series has 72 points.\
What I got from nnetar model in forecast in R that,

Average of 20 networks, each of which is\
a 16-8-1 network with 145 weights.

I got everything here that it is a 16-8-1 network i.e 16 nodes in input
layer and 8 nodes in hidden layer and 1 node in output layer.Here we use
16 lagged points of our time series.

My question is as nnetar gives different result in each run, in
production how can I decide to use the result of which run?Sometimes
result of 5th run is very nearer to my holdout point and sometimes
result of 1st run is very near.

Also the result varies for different seed.How to choose seed to get
accurate result.

Please answer although the question seems to be trivial.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-06-10 05:45:58
````
You need to train the model on as much data as you have available.\
neural networks use random starting values, so the output varies each
time you run it. Use a large number of starting values (set repeats to
be a large number) and the output variation will be small.\
The seed is not a parameter you can select to improve accuracy.

**Comment by robjhyndman** [comment]
```header
Originally posted by Reagan on 2017-06-12 13:16:43
````
Hello,

I\'ve only been working with these functions for a little while now, but
I cannot find anywhere that gives a good explanation of what the decay
part of the avNNet command does. Can anyone help explain? Thanks!

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-06-12 22:02:42
````
It is in the nnet::nnet command, not the avNNet command. See
[https://stackoverflow.com/a\...](https://stackoverflow.com/a/9575380/144157)
for an explanation.

**Comment by robjhyndman** [comment]
```header
Originally posted by Federico Botta on 2017-06-19 17:26:13
````
Hello,

Thanks for this post and the incredible forecast package!

I am having issues with the prediction intervals feature. I can
reproduce the examples on this page perfectly; however, when using my
own data set i obtain prediction intervals that do not contain the
forecasted point (for instance, the forecast returns a value of 457,053
and i get a prediction interval from 387,303 to 416,873).

This seems to happen in particular when i include an external regressor
in my model.

Interestingly, if I simulate the prediction interval using the
simulate() function as described above, and repeat that 1000 times, I
get a distribution perfectly centred around the estimate, and from there
i can get a perfectly fine interval.

I was wondering if anyone else has been having a similar issue?

For reference, the fitted network comes out as NNAR(17,1,10)\[12\], and
there\'s 157 observations in the original time series and the external
regressor (happy to provide the two data sets if needed).

Many thanks!

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-06-19 23:15:35
````
Maybe there\'s a problem when a covariate is used. Please post a minimal
reproducible example at
[https://github.com/robjhynd\...](https://github.com/robjhyndman/forecast/issues)

**Comment by robjhyndman** [comment]
```header
Originally posted by davidreilly7 on 2017-06-26 02:08:00
````
Hi Rob,

Thank you for this post. In the past I have had success with Neural Nets
used as a tool to forecast production tonnage (which was then used to
schedule trains for loading). I am however surprised by the relatively
poor performance of nnetar on the M3 data. Is that because much of M3 is
small sample size?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-06-26 04:22:24
````
Yes. ANNs need a large amount of data to work well, and the M3 time
series are quite short (anywhere between 14 and 126).

**Reply by robjhyndman** [reply]
```header
Originally posted by davidreilly7 on 2017-06-26 11:13:13
````
Ok, thanks. That\'s what I thought was the issue. You are so very
helpful - I really appreciate this!

**Comment by robjhyndman** [comment]
```header
Originally posted by Chhavi Garg on 2017-08-08 07:49:51
````
Hi Rob,\
I am using NNETAR to predict 2 days of electricity load. I have used 5
years of data to train the model with about 50 external regressors.
Training time is fairly high (about 1 hour on 8 GB RAM). I would like to
re-train the model, when about 2 more weeks of data has been
accumulated. This re-training will continue every time we have 2 more
weeks of data. Is their a way to adjust the weights without training the
model from scratch using NNETAR?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-08-08 07:53:50
````
No, but it is quite a simple function and you can easily modify it to
suit your purpose.

**Comment by robjhyndman** [comment]
```header
Originally posted by Neha Sharma on 2017-08-09 07:57:37
````
Hi Dr. Rob,

Great post!

I have a query. I am working on a real life problem of forecasting 3
days sales for a Retail store.\
I am thinking of applying a hybrid model(ARIMAX+Neural network) i.e.
Dynamic regression with regressors using auto.arima,then fitting Neural
network model on the residuals.The final forecast will be y= L+N where
L=forecast from ARIMAX and N= forecast of residuals from NNETAR. What do
you think of this approach?\
Also, I need your input on applying the cross validation techniques. I
have daily sales data from Jan14-June17. Would it be worth to tune the
parameters using cross validation techniques(Adding months/quarters) or
should I go ahead training the model only once (Let's say from
Jan14-Dec16) and measure the accuracy on the rest? (Test &amp;
Validation)?What could be the best approach as I need only 3 days
forecasts?

**Comment by robjhyndman** [comment]
```header
Originally posted by MHD K. on 2017-08-10 07:27:26
````
hello prof. Hyndman,

Please, Do you now mr. merwane roudane?\
He claims to be the one who wrote this R-code!\
Is he write this R-code? please tell me!

Thank you

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-08-10 07:33:45
````
Never heard of him. If you mean the forecast package for R, that was
written by me and others and has been on CRAN since 2006 and on my
personal website for a few years before that.

**Reply by robjhyndman** [reply]
```header
Originally posted by Dr. Mhd Fawaz Kourabi on 2017-08-12 21:55:26
````
He just claim that he wrote this exapmle in R.\
Thank you very much prof. Hayndman!

**Comment by robjhyndman** [comment]
```header
Originally posted by Steven Chau on 2017-08-18 09:52:44
````
Could you enlighten me how to get the lambda of .5 when
BoxCox.lambda(lynx) = 0.1521849?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-08-19 02:24:52
````
Chosen by eye.

**Comment by robjhyndman** [comment]
```header
Originally posted by farrukh aziz on 2017-08-30 12:56:44
````
[https://uploads.disquscdn.c\...](https://uploads.disquscdn.com/images/52676431f51e7fb0ba2e77fe9e8f0e304f8fecfc3ed33856469c8ea49a68edc2.png)

I am running nnetar on my data as nnetar(datasetname) i.e. passing no
parameters and the output which i am getting has lags=29, Can you please
help me in understanding what is \'lags\' here.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-08-30 20:47:00
````
Read
[https://www.otexts.org/fpp/9/3](https://www.otexts.org/fpp/9/3)

**Reply by robjhyndman** [reply]
```header
Originally posted by farrukh aziz on 2017-09-11 11:52:08
````
Hi Prof, i am trying to fix the values for decay, repeat and maxit, as
default value of decay = 0 which i have read may over fit the model,
please suggest the way forward to select these parameters.

**Comment by robjhyndman** [comment]
```header
Originally posted by Steven Chau on 2017-09-25 09:39:57
````
One thing I am not sure - the sigmoid is chosen as the activation
function and having the output range \[0,1\] However, the input layer is
scaled to its mean so would it make more sense to use tanh instead which
has mean of zero instead of 0.5?

**Comment by robjhyndman** [comment]
```header
Originally posted by Daniel  on 2017-10-11 16:59:40
````
Hi Rob,

Could you tell me which is the activation function in the hidden layer
in the nnetar function.

Thanks,\
Daniel

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-10-11 21:06:41
````
logistic. nnetar uses nnet::nnet which is documented in Ripley and
Venables MASS book.

**Comment by robjhyndman** [comment]
```header
Originally posted by Daniel Zapata on 2017-10-11 17:00:56
````
Hi rob,

Could you tell me which is the activation function on the hidden layer
in the nnetar function.

Thanks

**Comment by robjhyndman** [comment]
```header
Originally posted by Joshua Duncan on 2017-11-20 20:34:18
````
Hi Rob, thanks for your work. I\'m a little confused as to how the
nnetar model is non-linear. If there is only one hidden layer and no
non-linearity functions like sigmoid, relu, etc. how is this a
non-linear model? I imagine I\'m mistaken here as I\'m not an expert but
I am curious what you think. Thanks!

**Reply by robjhyndman** [reply]
```header
Originally posted by Joshua Duncan on 2017-11-20 21:03:42
````
I just noticed the comment above mine asking you what activation
function is being used. So it appears you do have a sigmoid
non-linearity function in the nnetar model. I believe that answers my
question unless you have something else to add. Thanks.

**Comment by robjhyndman** [comment]
```header
Originally posted by Marianna Wilson on 2018-05-18 08:34:54
````
Hi! Thank you for sharing this data.

**Comment by robjhyndman** [comment]
```header
Originally posted by Alex S on 2018-07-31 22:20:51
````
Prof. Hyndman, is there a reason why you went with feed forward neural
networks and not recurrent neural networks? The current prevailing
wisdom on the internet seems to be that RNN, and especially LSTM are the
best suited neural network models for time series forecasting?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2018-07-31 22:31:39
````
No reason at all. It was simply a proof of concept, not intended to be a
serious function.

**Comment by robjhyndman** [comment]
```header
Originally posted by Pam Grant on 2018-08-22 14:13:09
````
Thanks for sharing this data. Very interesting.

**Comment by robjhyndman** [comment]
```header
Originally posted by Rajkumar@Ambattur on 2019-04-15 09:25:04
````
Hi Rob,

Thanks for your blog.

I\'ve done a time series forecasting for my sales data using different
types of models(ets,arima,holtwinter,NNETAR ) .

I\'m also comparing the forecasted sales values with actual sales . Is
this the right way of doing it? and only NNETAR &amp; Holtwinter model are
forecasting nearer to the actual values.

Also, Can you tell as how to improve accuracy in NNETAR model?

Thanks in advance,\
Raj

**Comment by robjhyndman** [comment]
```header
Originally posted on 2017-05-29 23:25:09
````
No. The network is averaged over various random initializations. The
randomness in the simulations comes entirely from the error process.
